core-bundle/src/Resources/contao/classes/DataContainer.php

```diff
@@ -797,17 +797,23 @@ protected function generateButtons($arrRow, $strTable, $arrRootIds=array(), $bln
 				$attributes = ' class="' . $k . '"' . $attributes;
 			}
 
+			$href = $v['href'] ?? '';
+			if (!empty($v['route']))
+			{
+				$href = System::getContainer()->get('router')->generate($v['route'], array('id' => $arrRow['id']));
+			}
+
 			// Call a custom function instead of using the default button
 			if (\is_array($v['button_callback'] ?? null))
 			{
 				$this->import($v['button_callback'][0]);
-				$return .= $this->{$v['button_callback'][0]}->{$v['button_callback'][1]}($arrRow, $v['href'] ?? null, $label, $title, $v['icon'] ?? null, $attributes, $strTable, $arrRootIds, $arrChildRecordIds, $blnCircularReference, $strPrevious, $strNext, $this);
+				$return .= $this->{$v['button_callback'][0]}->{$v['button_callback'][1]}($arrRow, $href, $label, $title, $v['icon'] ?? null, $attributes, $strTable, $arrRootIds, $arrChildRecordIds, $blnCircularReference, $strPrevious, $strNext, $this);
```

Comment:
The 2nd parameter of the `button_callback` (href) will now never be `null`, right? 

Are there cases we would want to transform this (`$href ?: null`)? If not we should probably adjust the [documentation](https://docs.contao.org/dev/reference/dca/callbacks/#list-global-operations-operation-button).

Reply:
Fixed in 45758bc2d3e70cde3dc26c4883220424715e7f29.